src/main/java/net/minestom/server/inventory/InventoryImpl.java

```diff
@@ -0,0 +1,309 @@
+
+    private final int size;
+    protected final ItemStack[] itemStacks;
+
+    protected final ClickPreprocessor clickPreprocessor = new ClickPreprocessor(this);
+
+    private final TagHandler tagHandler = TagHandler.newHandler();
+
+    // the players currently viewing this inventory
+    protected final Set<Player> viewers = new CopyOnWriteArraySet<>();
+    protected final Set<Player> unmodifiableViewers = Collections.unmodifiableSet(viewers);
+    // (player -> cursor item) map, used by the click listeners
+    protected final Map<Player, ItemStack> cursorPlayersItem = new ConcurrentHashMap<>();
+
+    public static final @NotNull ClickHandler DEFAULT_HANDLER = new StandardClickHandler(
+            (builder, item, slot) -> slot >= builder.clickedInventory().getSize() ?
+                    IntIterators.fromTo(0, builder.clickedInventory().getSize()) :
+                    PlayerInventory.getInnerShiftClickSlots(builder, item, slot),
+            (builder, item, slot) -> IntIterators.concat(
+                    IntIterators.fromTo(0, builder.clickedInventory().getSize()),
+                    PlayerInventory.getInnerDoubleClickSlots(builder, item, slot)
+            ));
+
+    protected InventoryImpl(int size) {
+        this.size = size;
+        this.itemStacks = new ItemStack[getSize()];
+        Arrays.fill(itemStacks, ItemStack.AIR);
+    }
+
+    @Override
+    public int getSize() {
+        return size;
+    }
+
+    @Override
+    public byte getWindowId() {
+        return 1;
+    }
+
+    @Override
+    public @NotNull TagHandler tagHandler() {
+        return tagHandler;
+    }
+
+    @Override
+    public @NotNull ClickPreprocessor getClickPreprocessor() {
+        return clickPreprocessor;
+    }
+
+    @Override
+    public @NotNull ItemStack getCursorItem(@NotNull Player player) {
+        return cursorPlayersItem.getOrDefault(player, ItemStack.AIR);
+    }
+
+    @Override
+    public void setCursorItem(@NotNull Player player, @NotNull ItemStack cursorItem) {
+        updateCursor(player, cursorItem);
+        if (!cursorItem.isAir()) {
+            this.cursorPlayersItem.put(player, cursorItem);
+        } else {
+            this.cursorPlayersItem.remove(player);
+        }
+    }
+
+    @Override
+    public @NotNull Set<Player> getViewers() {
+        return unmodifiableViewers;
+    }
+
+    @Override
+    public boolean addViewer(@NotNull Player player) {
+        if (!this.viewers.add(player)) return false;
+
+        handleOpen(player);
+        return true;
+    }
+
+    @Override
+    public boolean removeViewer(@NotNull Player player) {
+        if (!this.viewers.remove(player)) return false;
+
+        handleClose(player);
+        return true;
+    }
+
+    /**
+     * Handles when a player opens this inventory, without actually updating viewers.
+     *
+     * @param player the player opening this inventory
+     */
+    protected void handleOpen(@NotNull Player player) {
```

Comment:
Now that I am further through the diff I will re-comment on this. I still don't understand the reason for this to be split from adding a viewer. it seems like a recipe to create bad behavior using this directly. Unless there is some major use case for this I think it should be merged into addViewer.